### PR TITLE
Added workaround to be able to skip a poll when the build is busy.

### PR DESCRIPTION
**Pull request by itavero**
I noticed that no other jobs where being polled when one was already running.
This was caused by the fact that the poll action for one of the jobs could not acquire a workspace because the build was in progress.
Unfortunately `AbstractProject` does not tell you which project it wants to poll when it calls `requiresWorkspaceForPolling()`.
I've added a workaround by using an `SCMPollListener` that stores the instance of the `AbstractProject`, so I can skip polling in `requiresWorkspaceForPolling()` and `compareRemoteRevisionWith()` when a build is in progress.

### CODE REVIEW

**Comment by rym002** [comment]
Would this be thread safe when many jobs are running?

**Comment by itavero** [comment]
@rym002 I was also thinking that this might be an issues, but then again, if Jenkins CI would do polling in parallel my problem would not exist in the first place.
But still, you might be right...

**Comment by rym002** [comment]
What about making the project an instance variable? This way it's not global and potentially impacting all projects.

**Comment by itavero** [comment]
@rym002 I do not quite understand what you mean. Can you elaborate?

**Comment by itavero** [comment]
@jsherwood-accurev @rym002 Anyone still maintaining this plug-in?

File: src/main/java/hudson/plugins/accurev/AccurevSCM.java [hunk 1]
```diff
@@ -403,6 +418,9 @@ public SCMRevisionState calcRevisionsFromBuild(AbstractBuild<?, ?> ab, Launcher
 
     @Override
     protected PollingResult compareRemoteRevisionWith(AbstractProject<?, ?> project, Launcher launcher, FilePath workspace, TaskListener listener, SCMRevisionState scmrs) throws IOException, InterruptedException {
+        if(skipNextPoll) {
```
Comment (rym002) [inline src/main/java/hudson/plugins/accurev/AccurevSCM.java]: can change to project.isBuilding(), set to this.project

File: src/main/java/hudson/plugins/accurev/AccurevSCM.java [hunk 2]
```diff
@@ -365,9 +366,23 @@ public ChangeLogParser createChangeLogParser() {
...
+        final boolean needSlaveForPolling = !DESCRIPTOR.isPollOnMaster() || accurevMode.isRequiresWorkspace();
+        if(!needSlaveForPolling) {
+            return needSlaveForPolling;
+        }
+
+        skipNextPoll = false;
+        if (AccuRevSCMPollListener.getLastProject() != null) {
+            AbstractProject<?, ?> project = AccuRevSCMPollListener.getLastProject();
+            AccuRevSCMPollListener.clearLastProject();
+            if(project.isBuilding()) {
```
Comment (rym002) [inline src/main/java/hudson/plugins/accurev/AccurevSCM.java]: @itavero add project!=null

File: src/main/java/hudson/plugins/accurev/AccurevSCM.java [hunk 3]
```diff
@@ -365,9 +366,23 @@ public ChangeLogParser createChangeLogParser() {
...
-        return accurevMode.isRequiresWorkspace() || needSlaveForPolling;
+        final AccurevMode accurevMode  = AccurevMode.findMode(this);
+        final boolean needSlaveForPolling = !DESCRIPTOR.isPollOnMaster() || accurevMode.isRequiresWorkspace();
+        if(!needSlaveForPolling) {
+            return needSlaveForPolling;
+        }
+
+        skipNextPoll = false;
+        if (AccuRevSCMPollListener.getLastProject() != null) {
+            AbstractProject<?, ?> project = AccuRevSCMPollListener.getLastProject();
```
Comment (rym002) [inline src/main/java/hudson/plugins/accurev/AccurevSCM.java]: @itavero Make AbstractProject<?, ?> project an instance variable

File: src/main/java/hudson/plugins/accurev/AccurevSCM.java [hunk 4]
```diff
@@ -365,9 +366,23 @@ public ChangeLogParser createChangeLogParser() {
...
-        AccurevMode accurevMode  = AccurevMode.findMode(this);
-        return accurevMode.isRequiresWorkspace() || needSlaveForPolling;
+        final AccurevMode accurevMode  = AccurevMode.findMode(this);
+        final boolean needSlaveForPolling = !DESCRIPTOR.isPollOnMaster() || accurevMode.isRequiresWorkspace();
+        if(!needSlaveForPolling) {
+            return needSlaveForPolling;
+        }
+
+        skipNextPoll = false;
+        if (AccuRevSCMPollListener.getLastProject() != null) {
```
Comment (rym002) [inline src/main/java/hudson/plugins/accurev/AccurevSCM.java]: @itavero  Remove AccuRevSCMPollListener

**Comment by rym002** [comment]
@itavero Please see my comments inline.

**Comment by itavero** [comment]
@rym002 I'll have a look at your proposals.

**Comment by itavero** [comment]
Squashed all changes into a single commit.

**Comment by casz** [comment]
@itavero Looks good to me, I would like to give @rym002 a chance to respond.

**Comment by rym002** [comment]
looks good to me also.

**Comment by itavero** [comment]
@casz Can you also publish a new version to the Jenkins CI plug-in repo?

**Comment by casz** [comment]
I did publish it :+1:

**Comment by itavero** [comment]
@casz Thanks. Unfortunately the plug-in is no longer detecting changes on our Jenkins CI 1.651.3 installation.
When I use my development build HPI it does work, so I'm guessing it has to do with more recent changes (so probably PR #25).
If I have some time left, I'll try to investigate this problem further.

**Comment by casz** [comment]
@itavero my testing shows, that polling works fine on a workspace build, but not on "Neither" job. Should have tested it more thoroughly.

**Comment by itavero** [comment]
We are using workspaces.
Op di 30 aug. 2016 om 15:06 schreef Joseph <notifications@github.com>

**Comment by casz** [comment]
@itavero I setup some logging, and it was constantly hitting activeproject not being null so that seems to be the reason for it detecting no changes.

**Comment by itavero** [comment]
Probably has to do with the conditions being combined in `requiresWorkspaceForPolling()` in `AccurevSCM`.
I think it should always be restored to `NULL` unless `(activeProject != null && activeProject.isBuilding())` is true.

**Comment by casz** [comment]
Just some logging example
These jobs are only polling, so building is done quickly :+1: 

```
Aug 30, 2016 4:13:00 PM INFO hudson.plugins.accurev.AccurevSCM compareRemoteRevisionWith
New active project
Aug 30, 2016 4:13:00 PM INFO hudson.plugins.accurev.AccurevSCM compareRemoteRevisionWith
New active project
Aug 30, 2016 4:14:00 PM INFO hudson.plugins.accurev.AccurevSCM compareRemoteRevisionWith
Workspace prevention
Aug 30, 2016 4:14:00 PM INFO hudson.plugins.accurev.AccurevSCM compareRemoteRevisionWith
Workspace prevention
Aug 30, 2016 4:16:00 PM INFO hudson.plugins.accurev.AccurevSCM compareRemoteRevisionWith
Workspace prevention
Aug 30, 2016 4:16:00 PM INFO hudson.plugins.accurev.AccurevSCM compareRemoteRevisionWith
Workspace prevention
```

**Comment by casz** [comment]
Question, would it not be better to override compareRemoteRevisionWith inside WorkspaceDeletegate?

**Comment by itavero** [comment]
If I remember correctly, this is not possible because Jenkins CI calls `requiresWorkspaceForPolling` before that method and in between it has a loop to simply keep trying to acquire the workspace.
And that will not work while the job is still building.

I tried to fix this in https://github.com/itavero/accurev-plugin/tree/fix-skip-poll
Currently checking if this improves things.

**Comment by casz** [comment]
@itavero seems to fix it. :+1:

Made a job with sleep 5 and polling seems to be working :+1:

**Comment by itavero** [comment]
Cool. I'll check the servers at my client tonight to see if everything is still running OK.
If so, I'll put in a PR.

**Comment by itavero** [comment]
@casz I'm using the build from the branch I told you about earlier, with workspaces for different jobs.
On Jenkins 1.651.3 this is causing builds to keep being triggered, because `Relocation required triggering build`.
Still something wrong, but I'm rather confident that it is not caused by my changes.

**Comment by itavero** [comment]
I've doubled checked in the AccuRev client and all the workspaces are located in the right place.
Will see if I can add some logging.

**Comment by itavero** [comment]
For some reason the polling is no longer performed on the slave, but on the master instead.
That's why Jenkins CI thinks a relocation is required, but because it are two different hosts, it can not perform the relocation.

The real question is, why is it trying to poll on the master instead of the slave, like it used to do...

**Comment by itavero** [comment]
Configuration item "Poll on master" is NOT checked.
So, I would assume that this is correct (even toggled it, behavior stays the same).

**Comment by casz** [comment]
I will debug the workspace bug but it is double folded I hit another bad one yesterday.

Knew about this performUpdate could be possible null pointer.
```
Getting a list of workspaces...
[workspacepolltest] $ /Applications/AccuRev/bin/accurev show -H *snip*:5050 -fx -p experiments wspaces
[workspacepolltest] $ /Applications/AccuRev/bin/accurev show -H *snip*:5050 -fx -p experiments streams
[workspacepolltest] $ /Applications/AccuRev/bin/accurev update -H *snip*:5050 -fx
FATAL: <?xml version="1.0" encoding="utf-8"?>
<acResponse
    command="update"
    TaskId="1737"/>
You are not in a directory associated with a workspace
FATAL: Failed authentication. (failed with exit code 1 )
FATAL: null
java.lang.NullPointerException
	at hudson.plugins.accurev.cmd.Update.performUpdate(Update.java:111)
	at hudson.plugins.accurev.delegates.ReftreeDelegate.doUpdate(ReftreeDelegate.java:148)
	at hudson.plugins.accurev.delegates.ReftreeDelegate.checkout(ReftreeDelegate.java:142)
	at hudson.plugins.accurev.delegates.AbstractModeDelegate.checkout(AbstractModeDelegate.java:191)
	at hudson.plugins.accurev.AccurevSCM.checkout(AccurevSCM.java:338)
	at hudson.scm.SCM.checkout(SCM.java:485)
	at hudson.model.AbstractProject.checkout(AbstractProject.java:1269)
	at hudson.model.AbstractBuild$AbstractBuildExecution.defaultCheckout(AbstractBuild.java:604)
	at jenkins.scm.SCMCheckoutStrategy.checkout(SCMCheckoutStrategy.java:86)
	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:529)
	at hudson.model.Run.execute(Run.java:1741)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:98)
	at hudson.model.Executor.run(Executor.java:410)
Finished: FAILURE
```

**Comment by itavero** [comment]
What parameter exactly is reported to be NULL?

**Comment by casz** [comment]
That I would have to find out :smile:

**Comment by casz** [comment]
It was a case of hostname was resolved to IP address which Accurev does not recognize when calling chws with -m.

So it was never located at the actual workspace which was on another machine.

**Comment by casz** [comment]
With polling on master and workspace on build and build is restricted to a certain machine. It just starts going in a endless loop of changing the workspace.

**Comment by itavero** [comment]
When polling on master is selected and a workspace/reference tree is used, but it is not on the master, I think we have two options:
- Ignore the setting and poll on the slave anyway (most efficient probably).
- Use the same mechanism for polling as is used when not using a workspace/reference tree (least efficient and less accurate).

There are probably valid arguments for both options. Personally I would prefer the first option.

**Pull request merged**